File: sequences/infinite-tribonacci.js
'use strict';

function InfiniteTribonacci() {
    this.index = 0;
    this._isFirst = true;
    this._currentSequence = [0, 1, 0, 2];
    this._prevSequence = [0, 1];
    this._prevPrevSequence = [0];
}

InfiniteTribonacci.prototype.next = function () {
    const value = this._currentSequence[this.index];

    this.index++;

    if (this.index === this._currentSequence.length) {
        this.index = 0;

        const current = [].concat(
            this._prevSequence,
            this._prevPrevSequence
        );

        this._prevPrevSequence = this._prevSequence;
        this._prevSequence = this._isFirst ? this._currentSequence : [].concat(
            this._prevSequence,
            this._currentSequence
        );
        this._currentSequence = current;

        this._isFirst = false;
    }

    return value;
};

module.exports = InfiniteTribonacci;

File: python/sequences/infinite_tribonacci.py
'use strict';

class InfiniteTribonacci:
    def __init__(self):
        self.index = 0
        self._isFirst = True
        self._currentSequence = [0, 1, 0, 2]
        self._prevSequence = [0, 1]
        self._prevPrevSequence = [0]

    def next(self):
        value = self._currentSequence[self.index]

        self.index += 1

        if self.index == len(self._currentSequence):
            self.index = 0

            current = self._prevSequence + self._prevPrevSequence

            self._prevPrevSequence = self._prevSequence

            if self._isFirst:
                self._prevSequence = self._currentSequence
            else:
                self._prevSequence = self._prevSequence + self._currentSequence

            self._currentSequence = current

            self._isFirst = False

        return value
